lib/std/target.zig

```diff
@@ -550,6 +550,103 @@ pub const Target = struct {
         }
     };
 
+    /// processor specific ABI
+    pub const TargetAbi = enum {
+        //TODO add ARM, Mips, and PowerPC
+        ilp32,
+        ilp32d,
+        ilp32e,
+        ilp32f,
+        lp64,
+        lp64d,
+        lp64f,
+
+        const Riscv32ABI = struct {
+            fn default(features: Cpu.Feature.Set) TargetAbi {
+                if (riscv.featureSetHas(features, .d)) {
+                    return .ilp32d;
+                } else if (riscv.featureSetHas(features, .e)) {
+                    return .ilp32e;
+                } else {
+                    return .ilp32;
+                }
+            }
+            fn validate(target_abi: TargetAbi, features: Cpu.Feature.Set) ParseError!void {
+                const has_e = riscv.featureSetHas(features, .e);
+                const has_f = riscv.featureSetHas(features, .f);
+                const has_d = riscv.featureSetHas(features, .d);
+                return switch (target_abi) {
+                    .ilp32e => if (has_e) void{} else error.FeatureAbiMismatch,
+                    .ilp32 => if (!has_e) void{} else error.FeatureAbiMismatch,
+                    .ilp32f => if (!has_e and has_f) void{} else error.FeatureAbiMismatch,
+                    .ilp32d => if (!has_e and has_d) void{} else error.FeatureAbiMismatch,
+                    else => error.ArchAbiMismatch,
+                };
+            }
+        };
+        const Riscv64ABI = struct {
+            fn default(features: Cpu.Feature.Set) TargetAbi {
+                if (riscv.featureSetHas(features, .d)) {
+                    return .lp64d;
+                } else {
+                    return .lp64;
+                }
+            }
+            fn validate(target_abi: TargetAbi, features: Cpu.Feature.Set) ParseError!void {
+                const has_f = riscv.featureSetHas(features, .f);
+                const has_d = riscv.featureSetHas(features, .d);
+                return switch (target_abi) {
+                    .lp64 => void{},
+                    .lp64f => if (has_f) void{} else error.FeatureAbiMismatch,
+                    .lp64d => if (has_d) void{} else error.FeatureAbiMismatch,
+                    else => error.ArchAbiMismatch,
+                };
+            }
+        };
+        pub fn default(arch: Cpu.Arch, features: Cpu.Feature.Set) ?TargetAbi {
+            return switch (arch) {
+                .riscv32 => Riscv32ABI.default(features),
+                .riscv64 => Riscv64ABI.default(features),
+                else => null,
+            };
+        }
+
+        // re-implementing @tagName because @tagName is not null-terminated in the stage0 compiler
```

Comment:
That's not true? `@tagName` returns a zero-terminated slice and that's been the case for quite some time.

Reply:
In the final output `zig` compiler from running `make`, when that zig compiler compiles `@tagName` it is null terminated. However, when I tried using `@tagName` I found that the strings were not null terminated when I compiled `zig`. So I assumed it was a difference between the intermediate compiler and the final compiler.

I'll try switching it back and trying again though. I'll post an update.

Reply:
I tried again, I still have the same issue with `@tagName`.

When I try to compile, eg with `zig build-obj -target riscv32-freestanding main.zig -mabi ilp32d` I get the following warning from LLVM:
```
'ilp32dilp32eilp32flp64dlp64f' is not a recognized ABI for this target (ignoring target-abi)
```

Reply:
Adding some detail from my memory of a few days ago, I believe I had tried putting in an
```zig
assert(tag_name_result[tag_name_result.len] == 0);
```
into TargetAbi.string, and that was OK on my personal zig code, but raised an exception when the compiler used it.

Reply:
Try with this patch:

```patch
diff --git a/lib/std/build.zig b/lib/std/build.zig
index 621499f8c..01d856bd2 100644
--- a/lib/std/build.zig
+++ b/lib/std/build.zig
@@ -2434,7 +2434,7 @@ pub const LibExeObjStep = struct {
         }
 
         if (self.target_abi) |target_abi| {
-            try zig_args.append(builder.fmt("-mabi={s}", .{target_abi.string()}));
+            try zig_args.append(builder.fmt("-mabi={s}", .{@tagName(target_abi)}));
         }
         if (self.code_model != .default) {
             try zig_args.append("-mcmodel");
diff --git a/lib/std/target.zig b/lib/std/target.zig
index 61a9a42e4..9a1557d35 100644
--- a/lib/std/target.zig
+++ b/lib/std/target.zig
@@ -611,25 +611,6 @@ pub const Target = struct {
             };
         }
 
-        // re-implementing @tagName because @tagName is not null-terminated in the stage0 compiler
-        fn tagName(value: anytype) [:0]const u8 {
-            const T = @TypeOf(value);
-            const info = @typeInfo(T);
-
-            switch (info) {
-                .Enum => |enum_info| {
-                    comptime var names: [enum_info.fields.len][:0]const u8 = undefined;
-                    inline for (enum_info.fields) |field, i| {
-                        names[i] = field.name ++ [0:0]u8{};
-                    }
-                    return names[@enumToInt(value)];
-                },
-                else => @compileError("tagName does not support type " ++ @typeName(T) ++ "."),
-            }
-        }
-        pub fn string(self: TargetAbi) [:0]const u8 {
-            return tagName(self);
-        }
         pub const ParseError = error{
             InvalidAbi,
             ArchAbiMismatch,
diff --git a/src/Compilation.zig b/src/Compilation.zig
index 81d99ae78..ef38b5585 100644
--- a/src/Compilation.zig
+++ b/src/Compilation.zig
@@ -1153,7 +1153,7 @@ pub fn create(gpa: *Allocator, options: InitOptions) !*Compilation {
         cache.hash.add(options.target.os.getVersionRange());
         cache.hash.add(options.is_native_os);
         cache.hash.add(options.target.abi);
-        cache.hash.addBytes(if (options.target_abi) |t| t.string() else ".");
+        cache.hash.addBytes(if (options.target_abi) |t| @tagName(t) else ".");
         cache.hash.add(ofmt);
         cache.hash.add(pic);
         cache.hash.add(pie);
@@ -3303,7 +3303,7 @@ pub fn addCCArgs(
             }
 
             if (comp.bin_file.options.target_abi) |target_abi| {
-                try argv.append(try std.fmt.allocPrint(arena, "-mabi={s}", .{target_abi.string()}));
+                try argv.append(try std.fmt.allocPrint(arena, "-mabi={s}", .{@tagName(target_abi)}));
             }
 
             if (comp.haveFramePointer()) {
@@ -3446,7 +3446,7 @@ pub fn addCCArgs(
             }
 
             if (comp.bin_file.options.target_abi) |target_abi| {
-                try argv.append(try std.fmt.allocPrint(arena, "-mabi={s}", .{target_abi.string()}));
+                try argv.append(try std.fmt.allocPrint(arena, "-mabi={s}", .{@tagName(target_abi)}));
             }
 
             if (target_util.clangAssemblerSupportsMcpuArg(target)) {
@@ -4370,7 +4370,7 @@ fn updateStage1Module(comp: *Compilation, main_progress_node: *std.Progress.Node
         .is_native_cpu = false, // Only true when bootstrapping the compiler.
         .llvm_cpu_name = if (target.cpu.model.llvm_name) |s| s.ptr else null,
         .llvm_cpu_features = comp.bin_file.options.llvm_cpu_features.?,
-        .llvm_target_abi = if (comp.bin_file.options.target_abi) |t| t.string().ptr else null,
+        .llvm_target_abi = if (comp.bin_file.options.target_abi) |t| @tagName(t) else null,
     };
 
     comp.stage1_cache_manifest = &man;
diff --git a/src/codegen/llvm.zig b/src/codegen/llvm.zig
index f26e474a7..bb55f2e1e 100644
--- a/src/codegen/llvm.zig
+++ b/src/codegen/llvm.zig
@@ -244,7 +244,7 @@ pub const Object = struct {
         // TODO handle float ABI better- it should depend on the ABI portion of std.Target
         const float_abi: llvm.ABIType = .Default;
 
-        const abi_name: ?[*:0]const u8 = if (options.target_abi) |t| t.string().ptr else null;
+        const abi_name: ?[*:0]const u8 = if (options.target_abi) |t| @tagName(t) else null;
 
         const target_machine = llvm.TargetMachine.create(
             target,
diff --git a/src/stage1/codegen.cpp b/src/stage1/codegen.cpp
index 54df7afbb..a3f8ce7e9 100644
--- a/src/stage1/codegen.cpp
+++ b/src/stage1/codegen.cpp
@@ -5399,8 +5399,9 @@ static LLVMValueRef get_enum_tag_name_function(CodeGen *g, ZigType *enum_type) {
     if (enum_type->data.enumeration.name_function)
         return enum_type->data.enumeration.name_function;
 
-    ZigType *u8_ptr_type = get_pointer_to_type_extra(g, g->builtin_types.entry_u8, false, false,
-            PtrLenUnknown, get_abi_alignment(g, g->builtin_types.entry_u8), 0, 0, false);
+    ZigType *u8_ptr_type = get_pointer_to_type_extra2(g, g->builtin_types.entry_u8, false, false,
+            PtrLenUnknown, get_abi_alignment(g, g->builtin_types.entry_u8), 0, 0, false,
+            VECTOR_INDEX_NONE, nullptr, g->intern.for_zero_byte());
     ZigType *u8_slice_type = get_slice_type(g, u8_ptr_type);
     ZigType *tag_int_type = enum_type->data.enumeration.tag_int_type;
 
@@ -5453,7 +5454,7 @@ static LLVMValueRef get_enum_tag_name_function(CodeGen *g, ZigType *enum_type) {
             continue;
         }
 
-        LLVMValueRef str_init = LLVMConstString(buf_ptr(name), (unsigned)buf_len(name), true);
+        LLVMValueRef str_init = LLVMConstString(buf_ptr(name), (unsigned)buf_len(name), false);
         LLVMValueRef str_global = LLVMAddGlobal(g->module, LLVMTypeOf(str_init), "");
         LLVMSetInitializer(str_global, str_init);
         LLVMSetLinkage(str_global, LLVMPrivateLinkage);
```

Reply:
That seems to work. Thanks!